x/auth/middleware/branch_store.go

```diff
@@ -0,0 +1,72 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/tendermint/tendermint/crypto/tmhash"
+)
+
+type branchStoreHandler struct {
+	next tx.Handler
+}
+
+// WithBranchedStore creates a new MultiStore branch and commits the store if the downstream
+// returned no error. It prohibits writes to the failed transactions.
+func WithBranchedStore(txh tx.Handler) tx.Handler {
+	return branchStoreHandler{next: txh}
+}
+
+// CheckTx implements tx.Handler.CheckTx method.
+// Do nothing during CheckTx.
+func (sh branchStoreHandler) CheckTx(ctx context.Context, req tx.Request, checkReq tx.RequestCheckTx) (tx.Response, tx.ResponseCheckTx, error) {
+	return sh.next.CheckTx(ctx, req, checkReq)
+}
+
+// DeliverTx implements tx.Handler.DeliverTx method.
+func (sh branchStoreHandler) DeliverTx(ctx context.Context, req tx.Request) (tx.Response, error) {
+	return branchAndRun(ctx, req, sh.next.DeliverTx)
+}
+
+// SimulateTx implements tx.Handler.SimulateTx method.
+func (sh branchStoreHandler) SimulateTx(ctx context.Context, req tx.Request) (tx.Response, error) {
+	return branchAndRun(ctx, req, sh.next.SimulateTx)
+}
+
+type nextFn func(ctx context.Context, req tx.Request) (tx.Response, error)
+
+// branchAndRun create a new Context based off of the existing Context with a MultiStore branch
+// in case message processing fails. At this point, the MultiStore
+// is a branch of a branch.
+func branchAndRun(ctx context.Context, req tx.Request, fn nextFn) (tx.Response, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	runMsgCtx, msCache := branchStore(sdkCtx, req.TxBytes)
+
+	rsp, err := fn(sdk.WrapSDKContext(runMsgCtx), req)
+	if err == nil {
+		// commit storage iff no error
+		msCache.Write()
+	}
+
+	return rsp, err
+}
+
+// branchStore returns a new context based off of the provided context with
+// a branched multi-store.
+func branchStore(sdkCtx sdk.Context, txBytes []byte) (sdk.Context, sdk.CacheMultiStore) {
+	ms := sdkCtx.MultiStore()
+	msCache := ms.CacheMultiStore()
+	if msCache.TracingEnabled() {
+		msCache = msCache.SetTracingContext(
+			sdk.TraceContext(
+				map[string]interface{}{
+					"txHash": fmt.Sprintf("%X", tmhash.Sum(txBytes)),
```

Comment:
use `tx.Hash()` for when the modular hashing ADR work gets merged on Tendermint

Reply:
done